pkg/ipcache/metadata.go

```diff
@@ -0,0 +1,406 @@
+	// Before allocating an identity, check if we should deallocate the old one
+	// first, if it exists.
+	if id, exists := IPIdentityCache.LookupByIP(prefix); exists {
+		// But not if it's only the kube-apiserver reserved identity. The
+		// kube-apiserver label can be associated with the host and CIDR
+		// labels.
+		realID := IdentityAllocator.LookupIdentityByID(context.TODO(), id.ID)
+		if realID != nil && !realID.Labels.Equals(labels.LabelKubeAPIServer) {
+			scopedLog := log.WithFields(logrus.Fields{
+				logfields.IPAddr:         prefix,
+				logfields.OldIdentity:    realID,
+				logfields.IdentityLabels: realID.Labels,
+			})
+
+			released, err := IdentityAllocator.Release(context.TODO(), realID)
+			if err != nil {
+				scopedLog.WithError(err).Warn(
+					"Failed to release previously assigned identity to IP, this might be a leak.",
+				)
+			} else {
+				scopedLog.WithFields(logrus.Fields{
+					"released":       released,
+					logfields.Labels: lbls,
+				}).Debug(
+					"Releasing old identity with previously assigned to IP and updating with new set of labels",
+				)
+			}
+		}
+	}
+
+	// If no other labels are associated with this IP, we assume that it's
+	// outside of the cluster and hence needs a CIDR identity. This might be a
+	// temporary identity.
+	if lbls.Equals(labels.LabelKubeAPIServer) {
+		// The release of the identitiy allocated has been handled above. This
+		// can happen if we discover that the prefix is also associated with
+		// other labels besides kube-apiserver, i.e. if the kube-apiserver is
+		// not running outside of the cluster.
+		return injectLabelsForCIDR(prefix, lbls)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), option.Config.IPAllocationTimeout)
+	defer cancel()
+	return IdentityAllocator.AllocateIdentity(ctx, lbls, false)
+}
+
+// injectLabelsForCIDR will allocate a CIDR identity for the given prefix. The
+// release of the identity must be managed by the caller.
+func injectLabelsForCIDR(prefix string, lbls labels.Labels) (*identity.Identity, bool, error) {
+	if !strings.Contains(prefix, "/") {
+		prefix = prefix + "/32"
+	}
+	_, cidr, err := net.ParseCIDR(prefix)
+	if err != nil {
+		return nil, false, err
+	}
+
+	allLbls := cidrlabels.GetCIDRLabels(cidr)
+	allLbls.MergeLabels(lbls)
+
+	log.WithFields(logrus.Fields{
+		logfields.CIDR:   cidr,
+		logfields.Labels: lbls, // omitting allLbls as CIDR labels would make this massive
+	}).Debug(
+		"Injecting CIDR labels for prefix",
+	)
+
+	return allocate(cidr, allLbls)
+}
+
+// FilterMetadataByLabels returns all the prefixes inside the IdentityMetadata
+// map which contain the given labels. Note that `filter` is a subset match,
+// not a full match.
+func FilterMetadataByLabels(filter labels.Labels) []string {
+	idMDMU.RLock()
+	defer idMDMU.RUnlock()
+	matching := make([]string, 0)
+	for prefix, lbls := range IdentityMetadata {
+		if bytes.Contains(lbls.SortedList(), filter.SortedList()) {
```

Comment:
`filter.SortedList()` should be done once before the looping to avoid having to sort it on every iteration.